### ISSUE TITLE
gym.Spaces conformance

### ISSUE THREAD

**Issue by ryanjulian** [rlworkgroup/akro#28]
This library would be more effective if it were phrased as an extension for `gym.Spaces`, i.e. akro types can be used as drop-in replacement types for `gym.Spaces` types, but with extra features.

This can probably be achieved by inheriting from spaces types and running the gym test suite on akro types in the CI

**Comment by ryanjulian** [comment]
Outline:
* Inherit akro types from `gym.Spaces` -- removing methods implemented by the `gym` superclasses 
* Remove `tf` and `theano` packages -- instead have helpers on the main package like `to_tf_tensor()` and `to_theano_tensor()`. You will have to guard these on whether the target frameworks are installed (since we can't require every user installs every framework). This can be done at import.
* Update tests and documentation
* Once everything is finished, bump the version so that a new version is sent to pip
* Refactor other rlworkgroup repos using akro to use the new interface, bumping their dependency version to the newest akro you just created.

**Comment by sud0nick** [comment]
To clarify the first point, each of the akro.Space subclasses (i.e. Box, Dict, etc) should inherit from gym.Space rather than akro.Space and none of the inherited methods should be overridden, correct?

**Comment by ryanjulian** [comment]
Yes. I think i should add some history so that this issue makes more sense.

The `spaces` library originated inside garage (then called rllab). Later it was copied and used for the `gym` library, and that implementation became overwhelmingly more popular. `garage.spaces` evolved for a bit and gained some nice features, and we eventually split it into `akro`. We believe there's a good reason to have a spaces library which is separate from `gym` (which is library of RL environments, not a toolbox or framework).

A lot of people have code written around `gym.spaces`, but `akro` has some features we don't want to give up. The proposal here is to make all of the `akro` spaces inherit from `gym.spaces`, so anybody can drop an `akro` space into existing code using `gym.spaces` and use the extra features from `akro`.

There has been a little API drift between `akro` and `gym`, so the first thing to do is to identify APIs on the `akro` spaces which exist on `gym.spaces` already, and just use the `gym` version (e.g. just delete that method and let it inherit from the gym superclass).

It's fine for the spaces to inherit from `akro.Space` as long as they still also inherit from their corresponding `gym` space, e.g.
```python
b = akro.Box()
assert isinstance(b, gym.spaces.Box)
```

`akro.Space` might become redundant and deserve deletion -- i'm not sure. How to organize the inheritance hierarchy I  think will become clear as you do the refactor.

**Comment by ryanjulian** [comment]
Once you have refactored the `akro` library, there are several `rlworkgroup` packages which rely on `akro` which you will have to visit and update with any API changes you made (e.g. renames, type checks, etc.).

**Comment by sud0nick** [comment]
Got it, thanks for the explanation.  Would you prefer I also move the akro spaces to their own spaces module similar to how gym.spaces is set up?  That way it's more in line with the structure of gym so we can import like so:

    from gym.spaces import Box
    from akro.spaces import Box

**Comment by ryanjulian** [comment]
hmm for now let's just keep it `akro.Box` because i don't expect `akro` to have anything other than these yet. during implementation we might find that the other style makes more sense and choose that.

**Comment by sud0nick** [comment]
Sounds good.  I've finished modifying the inheritance and want to push my new branch but it looks like I don't have rights yet.  Can you grant them?

Also, in your second point you mention the helper functions should be in the main package.  What are you considering to be the main package?  I can't find any file with a main() method.  Or does this mean to add the helpers in each instance of Space?

**Comment by ryanjulian** [comment]
sorry for not making my original comment clearer.

by "main" package I meant that we should remove the packages `akro.tf` and `akro.theano`. There are methods in those packages which are specialized for the frameworks, e.g. `akro.tf.Box.new_tensor_variable()`. I am proposing changing the pack of those methods to `akro.Box.to_tf_placeholder()`.

This has the caveat that, of course, you can't call these functions unless `tensorflow` is installed. We don't want people to have to install 2, 3, or 4 different giant numerical libraries just to use `akro`, so we won't make `tensorflow` a dependency. Instead, we can catch when someone tries to use a specialized function without the corresponding library and print a useful error.

e.g.
```python
def my_special_tf_function(self):
    try:
        import tensorflow as tf
        tf.do_some_stuff()
    except ImportError:
        raise RuntimeError('You must install tensorflow to use akro to create tensorflow tensors')
```

This style is tiresome and might be inefficient, so we can instead do all the checks at once by doing this in `__init__.py`

**akro/__init__.py**
```python
tf = False
try:
    import tensorflow as tf
except ImportError:
    pass

theano = False
try:
    import theano
except ImportError:
    pass

# future!
torch = False
try:
    import torch
except ImportError:
    pass
```

**akro/my_module.py**
```python
import akro

class MyType:

    def my_special_tf_function(self):
        if not akro.tf:
            raise RuntimeError('Helpful message')

        tf.do_some_stuff()
```

However this is also kind of cumbersome, so what if we made a decorator instead

**akro/requires.py**
```python
import akro

def requires_tf(func):
    def check(*args, **kwargs):
        if not akro.tf:
            raise RuntimeError('You must install tensorflow before you can use the function {}'.format(func))

        func(*args, **kwargs)
```

**akro/my_module.py**
```python
from akro.requires import requires_tf

class MyType:

   @requires_tf
    def my_special_tf_function(self):
        tf.do_some_stuff()
```

**Comment by sud0nick** [comment]
@ryanjulian I've come across a strange problem I'm hoping you can give some insight on.  Currently, when a Tuple is constructed it calls `c.new_tensor_variable('tmp', extra_dims=0).dtype for c in components`.  I've since changed the `new_tensor_variable` method in each class to `to_tf_placeholder` and added the code from the old `tf` version of the class which simply calls

    return akro.tf.placeholder(
            dtype=akro.tf.uint8,
            shape=[None] * extra_dims + [self.flat_dim],
            name=name)

This is failing in the CI process because it's returning a NoneType object.  However, I've verified that a Tensor object does get created successfully but for some unknown reason `None` is returned.  Regardless of this issue, looking at the existing code in the master branch the `new_tensor_variable` method raises a NotImplementedError.  How did this ever pass the CI tests and do you have any idea why I'm getting `None` when returning a Tensor?

**Comment by ryanjulian** [comment]
@krzentner please advise

**Comment by krzentner** [comment]
I have no idea what is causing this. According to the [tf.placeholder](https://www.tensorflow.org/api_docs/python/tf/placeholder) documentation, it should not ever return None. Have you tried stepping through it with a debugger?

If you're still having trouble with it, can you push up the change including the broken tests?

**Comment by sud0nick** [comment]
Thanks, I got it figured out.  The decorator I created wasn't returning the value of the `func(*args, **kwargs)` which is why I could see it be created inside the call to `to_tf_placeholder` but couldn't see it outside of that call. :)

**Comment by sud0nick** [comment]
@ryanjulian Can this issue be closed?

**Comment by sud0nick** [comment]
Resolved in PR https://github.com/rlworkgroup/akro/pull/34
